server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -718,6 +733,40 @@ public boolean isRequireAlias() {
         return requireAlias;
     }
 
+    @Override
+    public Index getConcreteWriteIndex(IndexAbstraction ia, Metadata metadata) {
+        if (opType() != DocWriteRequest.OpType.CREATE) {
+            return ia.getWriteIndex();
+        }
+
+        if (ia.getType() != IndexAbstraction.Type.DATA_STREAM) {
+            return ia.getWriteIndex();
+        }
+
+        DataStream dataStream = metadata.dataStreams().get(ia.getName());
+        if (dataStream.isTimeSeries(metadata::index) == false) {
+            return ia.getWriteIndex();
+        }
+
+        try (XContentParser parser = contentType.xContent().createParser(TS_EXTRACT_CONFIG, source().streamInput())) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+            ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.nextToken(), parser);
+            ensureExpectedToken(XContentParser.Token.VALUE_STRING, parser.nextToken(), parser);
+            String timestampAsString = parser.text();
+            // TODO: deal with nanos too here.
+            // (the index hasn't been resolved yet, keep track of timestamp field metadata at data stream level, so we can use it here)
+            Instant timestamp = Instant.from(DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parse(timestampAsString));
+
+            Index result = dataStream.selectWriteIndex(timestamp, metadata);
+            if (result == null) {
+                throw new IllegalArgumentException("no index available for a document with an @timestamp of [" + timestampAsString + "]");
```

Comment:
It would be great if we could add a bit more useful information here. I think I would rephrase it as "the document timestamp [2022-01-07T19:04:41Z] is outside of ranges of currently writable indices: [[2022-01-06T00:00:00.000Z-2022-01-06T16:02:12.251Z], [2022-01-06T16:02:12.251Z-2022-01-06T20:02:12.251Z]]" or something like this.

Reply:
Also implemented via: f97d88935098460c2f1a7cfafb6d4271636b60a5